crates/goose/src/token_counter.rs

```diff
@@ -68,16 +71,137 @@ impl TokenCounter {
         }
     }
 
-    pub fn count_tokens(&self, text: &str, model_name: Option<&str>) -> usize {
+    fn get_tokenizer(&self, model_name: Option<&str>) -> &Tokenizer {
         let tokenizer_key = Self::model_to_tokenizer_key(model_name);
-        dbg!(&model_name, &tokenizer_key);
-        let tokenizer = self
-            .tokenizers
+        self.tokenizers
             .get(tokenizer_key)
-            .expect("Tokenizer not found");
+            .expect("Tokenizer not found")
+    }
+
+    pub fn count_tokens(&self, text: &str, model_name: Option<&str>) -> usize {
+        let tokenizer = self.get_tokenizer(model_name);
         let encoding = tokenizer.encode(text, false).unwrap();
         encoding.len()
     }
+
+    fn count_tokens_for_tools(&self, tools: &[Tool], model_name: Option<&str>) -> usize {
```

Comment:
I wonder if this should be something the system exposes rather than Goose has to compute? since tools are coupled with systems and this is almost like part of the description of a system

Reply:
yea, we thought about this. It does make sense for the system to compute all this I agree, but it adds more dependency to them as well. they'll have to install/load tokenizers and run them. Maybe once MCP is further implemented we can use the sampling feature to get these directly onto the system side rather than having the agent do it? adds overhead though